### PR TITLE
302 auto outliers

### PR DESCRIPTION
**Pull request by AnneONS**
## Pull Request submission

Creation of functions to detect and apply automatically detected outliers.

This adapted code written by GZ using the pandas quantile function.

The script `auto_outliers.py` is called from `outlier_main.py` and: 
-  validates `clip_upper` and `clip_lower` parameters from the configuration
-  calculates the outliers based on the parameters supplied
- outputs the number of outliers in the outlier_val_col column, along with the total number of non-zero entries.

A unit test is given for the outlier calculation function.

NOTE: We're still waiting for confirmation about which column the outliers should be detected in, currently the outlier_val_col is set to '701' in the code, this column is the total expenditure for Civil in the short forms, the short form equivalent of q211.

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #302 and #297

#### Code

- [x] **Code runs** The code runs on my machine 
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [x] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [x] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [x] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [x] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [x] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [x] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/outlier_detection/auto_outliers.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import logging
+import pandas as pd
+
+OutlierMainLogger = logging.getLogger(__name__)
+
+
+def validate_config(upper_clip: float, lower_clip: float):
+    """Validate the outlier config settings.
+
+    args:
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py]: Capitalise the args to Args? Also Arguments or Args?

File: src/outlier_detection/auto_outliers.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    """
+    OutlierMainLogger.info(f"Upper clip: {upper_clip}, Lower clip: {lower_clip}")
+
+    if (upper_clip == 0) and (lower_clip == 0):
+        OutlierMainLogger.warning(
+            "upper_clip and lower_clip both zero: "
+            "All auto_outlier flags will be False"
+        )
+    elif (upper_clip < 0) | (lower_clip < 0):
+        OutlierMainLogger.error("upper_clip and lower_clip cannot be negative")
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py:33]: preference for a : or a -- for the args (-- is used elsewhere in pipeline e.g. validation.py)?

File: src/outlier_detection/auto_outliers.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+import pandas as pd
+
+OutlierMainLogger = logging.getLogger(__name__)
+
+
+def validate_config(upper_clip: float, lower_clip: float):
+    """Validate the outlier config settings.
+
+    args:
+        upper_clip (float): The percentage for upper clipping as float
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py:17]: Put types in the doc strings in () or not? ("df -- DataFrame containing contributor data" or df (pd.DataFrame))

File: src/outlier_detection/auto_outliers.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    if (upper_clip == 0) and (lower_clip == 0):
+        OutlierMainLogger.warning(
+            "upper_clip and lower_clip both zero: "
+            "All auto_outlier flags will be False"
+        )
+    elif (upper_clip < 0) | (lower_clip < 0):
+        OutlierMainLogger.error("upper_clip and lower_clip cannot be negative")
+        raise ImportError
+    elif upper_clip + lower_clip > 1.0:
+        OutlierMainLogger.error("upper_clip and lower_clip must sum to < 1")
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py:36]: <= within the error message?

Comment (smithv1) [reply src/outlier_detection/auto_outliers.py:36]: so change "OutlierMainLogger.error("upper_clip and lower_clip must sum to < 1")" 
to: "OutlierMainLogger.error("upper_clip and lower_clip must sum to <= 1")"?

File: src/outlier_detection/auto_outliers.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        raise ImportError
+
+
+def outlier_flagging(
+    df: pd.DataFrame,
+    upper_clip: float,
+    lower_clip: float,
+    groupby_cols: list = ["period, cellnumber"],
+    value_col: str = "211",
+):
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py]: add the return type hint and elsewhere where needed.

File: src/outlier_detection/auto_outliers.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        groupby_cols (list[str]): The columns to be grouped by
+        value_col (str): The name of the column outliers are calculated for
+    returns:
+        df (pd.DataFrame): The dataframe with a outlier flag column.
+    """
+    lower_band = 0 + lower_clip
+    upper_band = 1 - upper_clip
+
+    # Note: selectiontype=='P' doesn't exist in anonymised data!
+    filter_cond = (df[value_col] > 0) & (df.selectiontype=='P')
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py]: can we add the column name value_col to args? Also consistency check: do we prefer check df.column_name or df['column_name']?

File: src/outlier_detection/auto_outliers.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    filtered_df = df[filter_cond]
+
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py:76]: maybe add a comment here describe what next lines are doing and what interpolation="nearest" does

File: src/outlier_detection/auto_outliers.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    quantiles_up = (
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py:77]: 58 to 76: could this be one function called twice - what are the limits of functionising?

File: src/outlier_detection/auto_outliers.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        filtered_df[groupby_cols + [value_col]]
+        .groupby(groupby_cols)
+        .quantile([upper_band], interpolation="nearest")
+        .reset_index()
+    )
+    quantiles_up = quantiles_up.rename(columns={value_col: "upper_band"})[
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py:83]: not sure what the second half of this is doing when reading - could add comment.

File: src/outlier_detection/auto_outliers.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+def calc_num_outliers(df: pd.DataFrame, outlier_val_col: str):
+    """Calculate and number of outliers flagged.
+    
+    Also calculate the total number of non-zero entries in the 
+    outlier value column.
+
+    args:
+        df (pd.DataFrame): The dataframe used for finding outliers
+        value_col (str): The name of the column outliers are calculated for
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py]: change value_col to outlier_val_col

File: src/outlier_detection/auto_outliers.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    """Flag outliers for quantiles specified in the config.
+
+
+    Calculate the automatic outlier flag (Boolean) in a column
+    called auto_outlier.
+
+    Use values specified in the config file for the upper and lower
+    clipping values.
+
+    Notes:
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py:165]: Notes, Args, Returns calitalised here - consistency check

File: src/outlier_detection/auto_outliers.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        - The value column for the detection of outliers is assumed to be '211'.
+        - Outliering is not applied to 'census' data, so flags are only created
+            where 'seclectiontype' = 'P'.
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py]: 'seclectiontype' = 'P'. - add further explanation?

File: src/outlier_detection/auto_outliers.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+    Returns:
+        df (pd.DataFrame): The full dataset with flag column for outliers.
+    """
+    OutlierMainLogger.info("Starting outlier detection...")
+
+    # Validate the outlier configuration settings
+    validate_config(upper_clip, lower_clip)
+
+    # Specify the value column and the groupby columns for outliers
+    outlier_val_col = "709"
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py]: Maybe add outlier_val_col as an arg?

File: src/outlier_detection/auto_outliers.py [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+
+    df[outlier_val_col] = pd.to_numeric(df[outlier_val_col], errors='coerce')
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py]: Could add explanation of errors='coerce'

File: src/outlier_detection/auto_outliers.py [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    groupby_cols = ["period", "cellnumber"]
```
Comment (smithv1) [inline src/outlier_detection/auto_outliers.py]: Add groupby_cols as arg?

File: src/outlier_detection/outlier_main.py [hunk 16]
```diff
@@ -1 +1,56 @@
...
+    OutlierMainLogger.info("Starting Auto Outlier Detection...")
+
+    upper_clip = config["outliers"]["upper_clip"]
+    lower_clip = config["outliers"]["lower_clip"]
+    df_auto_flagged = auto.auto_clipping(df, upper_clip, lower_clip)
+
+    OutlierMainLogger.info("Finished Auto Outlier Detection.")
+
+    # output the outlier flags for QA
+    if config["global"]["network_or_hdfs"] == "network":
```
Comment (smithv1) [inline src/outlier_detection/outlier_main.py:48]: Do we not want this or similar to also run when set to hdfs

**Review by smithv1 — CHANGES_REQUESTED** [review]

File: src/outlier_detection/auto_outliers.py (hunk 5 above)
Comment (AnneONS) [reply src/outlier_detection/auto_outliers.py]: Good point, I forgot that!

File: src/outlier_detection/auto_outliers.py (hunk 12 above)
Comment (AnneONS) [reply src/outlier_detection/auto_outliers.py]: Yes, I can do this know that I know more.

File: src/outlier_detection/auto_outliers.py (hunk 4 above)
Comment (AnneONS) [reply src/outlier_detection/auto_outliers.py:36]: Thanks for spotting that, should have the logical condition upper_clip + lower_clip >= 1

File: src/outlier_detection/auto_outliers.py (hunk 13 above)
Comment (AnneONS) [reply src/outlier_detection/auto_outliers.py]: Good point, I'll look at doing this.

File: src/outlier_detection/outlier_main.py (hunk 16 above)
Comment (AnneONS) [reply src/outlier_detection/outlier_main.py:48]: At this stage not sure how we are outputting files in hdfs

**Review by smithv1 — APPROVED** [review]

**Comment by smithv1** [comment]
Code looks good, tests good and runs via main.py

**Review by jwestw — APPROVED** [review]
This looks really good. Excellent code @AnneONS

**Pull request merged**